feat: add `nulls` option to `createIndex`

jonahkagan opened the pull request:
Closes https://github.com/salsita/node-pg-migrate/issues/1470

Adds an optional `nulls` option to the `createIndex` function with two possible values: `'distinct'` and `'not distinct'`. If provided, adds a `NULLS DISTINCT` or `NULLS NOT DISTINCT` clause to the index creation command.

Note that in the original issue I quoted from some docs saying that there were two other nulls clause possibilities: `NULLS FIRST` and `NULLS LAST`, but those actually are a separate option that is part of the sort order specifier, so I didn't include them (since there's no option for sort order in `createIndex` yet).

brenoepics commented:
> I added the migration test file. It looks like there's an integration test failure in CI (looks like an outdated snapshot maybe?).
> 
> I haven't yet been able to figure out how to run the integration tests locally - I may be missing some dependencies since I'm getting the following error:
> 
> ```
> 
>  FAIL   integration  test/integration/db-config-it.spec.ts > node-pg-migrate config file and env fallback (PG 17)
> Error: Could not find a working container runtime strategy
>  ❯ getContainerRuntimeClient node_modules/testcontainers/src/container-runtime/clients/client.ts:63:9
>  ❯ PostgreSqlContainer.start node_modules/testcontainers/src/generic-container/generic-container.ts:87:20
>  ❯ PostgreSqlContainer.start node_modules/@testcontainers/postgresql/src/postgresql-container.ts:49:43
>  ❯ Module.setupPostgresDatabase test/integration/utils.ts:71:10
>      69|   databaseName: string = 'node_pg_migrate'
>      70| ): Promise<StartedPostgreSqlContainer> {
>      71|   return await new PostgreSqlContainer(containerImage)
>        |          ^
>      72|     .withUsername('ubuntu')
>      73|     .withPassword('ubuntu')
>  ❯ test/integration/db-config-it.spec.ts:36:21
> ```
> 
> If it's just a quick snapshot update it may be quicker for someone with a working development environment to do it rather than to try to help unblock me.

Hmm, maybe https://github.com/testcontainers/testcontainers-node/issues/913 or https://github.com/testcontainers/testcontainers-node/issues/793 ?

Are you using `nektos/act` or something like that? But yeah, I think it is related to snapshots.

jonahkagan commented:
@brenoepics thanks for offering! Looking at the snapshot diff you posted, I actually realized that the test migration was missing the `unique` option. I updated the migration and also added a guard to prevent that from happening again. See eaea60b for details.

If you wouldn't mind posting the result of the snapshot after that change, I can validate it before you push it. Thanks for offering to help!

brenoepics commented:
<img width="1701" height="425" alt="image" src="https://github.com/user-attachments/assets/671428aa-7624-4932-9996-5dc1cf408087" />

jonahkagan commented:
Hm, I still don't see the "NULLS" clause in the snapshot. I wonder if Postgres filters it out if you pass in the default value ("DISTINCT"). I'm gonna try changing it to "NOT DISTINCT": 5cd4572

brenoepics commented:
```sql
### MIGRATION 095_index_nulls (UP) ###
CREATE TABLE "t095" (
  "id" integer NOT NULL
);
CREATE UNIQUE INDEX "t095_id_unique_index" ON "t095" ("id");
INSERT INTO "public"."pgmigrations" (name, run_on) VALUES ('095_index_nulls', NOW());
```

jonahkagan commented:
Awesome! That looks great!

brenoepics commented:
Last thing, `nulls` on `createIndex` is not available on PG 13 and 14. See https://www.postgresql.org/docs/14/sql-createindex.html and https://www.postgresql.org/docs/17/sql-createindex.html

maybe we can do something like we did at https://github.com/salsita/node-pg-migrate/blob/14cd48a6bf2a9aaac78b4d24a5142e11ee583bf3/test/migrations/093_alter_column_expression.js

All other versions are working fine now.

jonahkagan commented:
Updated with a version guard. Thanks for the pointer to the example
